src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt

```diff
@@ -205,4 +319,54 @@ class ProjectService(
         }
         return projectMembersWithUsers.toGrpcProjectMembers()
     }
+
+    override suspend fun getProjectPaperById(request: Base.Id): GrpcProject.Paper {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+        val projectPaperId = parseUUID(request.id, EntityType.PROJECT_PAPER)
+        val projectPaper = projectPaperRepo.getProjectPaperById(projectPaperId)
+        val projectId = projectPaper.projectId
+        val projectMembers = projectMemberRepo.getProjectMembers(projectId)
+
+        if (!projectMembers.any { it.userId == currentUser.id }) {
+            verifyServerAdminRole(currentUser) {
+                throw UnauthorizedException.Single(
+                    EntityType.PROJECT,
+                    projectId.toString(),
+                    AccessType.READ,
+                    it,
+                )
+            }
+        }
+
+        val paper = paperRepo.getPaperById(projectPaper.paperId)
+        val authors = authorOfPaperTableRepo.getAuthorsOfPaperById(paper.id).map { it.toGrpcAuthor() }
+        val backwardReferences = citationTableRepo.getBackwardsReferencedPaperIdsOfPaperById(
+            paper.id,
+        ).map { it.toString() }
+        val reviews = reviewTableRepo.getAllReviewsForProjectPaper(projectPaperId)
+            .map {
+                val selectedCriteriaIds = reviewHasCriterionTableRepo.getSelectedCriteriaIdsForReviewById(it.id)
+                it.toGrpcReview(selectedCriteriaIds.map { criterion -> criterion.toString() })
+            }
+
+        return ProjectPaperWithPaper(projectPaper, paper).toGrpcProjectPaper(authors, backwardReferences, reviews)
+    }
+
+    override suspend fun getAllProjectPapersForProject(request: Base.Id): GrpcProjectPaper.List =
+        getProjectPapers(request)
+
+    override suspend fun getPapersToReviewForProject(request: Base.Id): GrpcProjectPaper.List {
+        val predicate: (ProjectPaperWithPaper, Map<ProjectPaper, List<GrpcReview>>, String) -> Boolean =
+            { projectPaper, projectPaperReviewsMap, currentUserId ->
+                val isAlreadyReviewedByCurrentUser = projectPaperReviewsMap[projectPaper.projectPaper]
+                    ?.any { review -> review.userId == currentUserId } == true
+                val isAlreadyDecided =
```

Comment:
```suggestion
                val isStillUndecided =
```
and when you use it as well. Because now you inverted the check

Reply:
True. Thanks.